Fix a whole bunch of deadlocks

MaxKellermann opened the pull request:
No matter where I look in the Kodi source code, I see thread-safety bugs. So many threads (way too many for my taste!), so little attention to thread-safety.

MartijnKaijser commented:
Please first against master branch as that is our main dev branch. You can keep this open for now.

FernetMenta commented on xbmc/epg/EpgContainer.cpp:
```diff
@@ -239,7 +239,17 @@ void CEpgContainer::LoadFromDB(void)
+         g_graphicsContext); note that ShowProgressDialog() is
+         sometimes called with m_critSection locked and sometimes
+         without; this is a major bug that must be addressed
+         eventually */
```
It will. I am working on a split of render thread from main thread. This should cure quite a few similar issues.

MaxKellermann replied:
On 2015/12/20 22:06, Rainer Hochecker <notifications@github.com> wrote:
> It will. I am working on a split of render thread from main thread. This should cure quite a few similar issues.

No, the split will not cure those issues.  It will just hide them.

MaxKellermann commented:
On 2015/12/20 21:51, Martijn Kaijser <notifications@github.com> wrote:
> Please first against master branch as that is our main dev branch. You can keep this open for now. 

Cherry-pick the commits if you want them in master.  I'm tired of
resubmitting my bug fixes over and over with no change (except for the
rebase).  I don't like how you manage the Kodi branches, because I
feel this is wasting too much (of my) time with duplicated work.

MartijnKaijser commented:
Then only submit them to master and let someone else do cherry picking. Like it or not but this is how we handle it

MaxKellermann committed:
pvr: fix a bunch of deadlock bugs

Unlock m_critSection whenever calling a global object's method.  This
is very important to avoid deadlocks, because the callees may take
another lock, which may lead to a deadlock.

One of these bugs was observed because JobWorker deadlocked inside
EPG::CEpgContainer::LoadFromDB(), waiting for g_graphicsContext; main
thread holding g_graphicsContext, waiting for
PVR::CPVRGUIInfo::m_critSection; PVRGUIInfo holding
PVR::CPVRGUIInfo::m_critSection, waiting for
CEpgContainer::m_critSection.  Circle closed.

There are many, many more deadlock bugs in the PVR code.  Mutexes are
used in a very inconsistent way; sometimes, thread-unsafe objects are
not locked at all, and sometimes, mutex are held for too long while
calling other functions that will lock another mutex, eventually
leading to a circular deadlock.

MaxKellermann committed:
epg/EpgContainer: fix deadlock bug in LoadFromDB()

This bug was observed because JobWorker deadlocked inside
EPG::CEpgContainer::LoadFromDB(), waiting for g_graphicsContext; main
thread holding g_graphicsContext, waiting for
PVR::CPVRGUIInfo::m_critSection; PVRGUIInfo holding
PVR::CPVRGUIInfo::m_critSection, waiting for
CEpgContainer::m_critSection.  Circle closed.

There are many, many more deadlock bugs in the PVR code.  Mutexes are
used in a very inconsistent way; sometimes, thread-unsafe objects are
not locked at all, and sometimes, mutex are held for too long while
calling other functions that will lock another mutex, eventually
leading to a circular deadlock.

MaxKellermann committed:
GUIWindowManager: protect m_idCache with g_graphicsContext

In GetWindow(), the CGUIWindowManagerIdCache::Get() call is
unprotected, while all other accesses in this class are protected with
g_graphicsContext.  Get() is not thread-safe, and must be protected.

MaxKellermann committed:
pvr/PVRManager: make m_pvrWindowIds "const"

The "const" clarifies that this is a read-only variable that does not
need mutex protection.

FernetMenta commented:
jenkins build this please

MartijnKaijser commented:
@MaxKellermann Thank you
